### PR TITLE
EOS-20858: Pacemaker Alert further log replication for >3nodes.

### PR DESCRIPTION
**Pull request by shindeam**
Signed-off-by: Amol Shinde <amol.shinde@seagate.com>

## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    https://jts.seagate.com/browse/EOS-20858
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Pacemaker Alert further log replication for >3nodes.
  </code>
</pre>
## Solution
<pre>
  <code>
   - Get all online corosync nodes
   - Get self nodeid
    - Find out way of identifying the highest node from online nodes
    - Send alert only from lowest or highest online node
    - Environment variable are available in self.crm_env
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
Logs on node3:
2021-07-21 07:57:49 alert_monitor INFO [__init__] AlertEventFilter initialized.
2021-07-21 07:57:49 alert_monitor INFO [process_alert] Processing event for NodeAlertMonitor
2021-07-21 07:57:50 alert_monitor INFO [get_node_id] List of online nodes ids in cluster in sorted ascending order: ['1', '2', '3']
2021-07-21 07:57:50 alert_monitor INFO [get_node_id] Local node name: srvnode-3.data.private
2021-07-21 07:57:50 alert_monitor INFO [process_alert] Before checking if local node is highest in cluster-:
2021-07-21 07:57:50 alert_monitor INFO [process_alert] Local node name: srvnode-3.data.private
2021-07-21 07:57:50 alert_monitor INFO [process_alert] All online nodes: ['1', '2', '3']
2021-07-21 07:57:50 alert_monitor INFO [process_alert] Last Nodes IDs: 3
2021-07-21 07:57:50 alert_monitor INFO [generate_iem] Sending an IEC: "IEC:IS0080010002:Server srvnode-1.data.private has joined back the cluster. System is restored. Extra Info: host=srvnode-1.data.private; status=member;" to syslog
2021-07-21 07:57:50 alert_monitor INFO [process_alert] Sent IEM alert from the node - name: srvnode-3.data.private
, id: 3

Logs on node2:
2021-07-21 07:57:50 alert_monitor INFO [get_node_id] List of online nodes ids in cluster in sorted ascending order: ['1', '2', '3']
2021-07-21 07:57:50 alert_monitor INFO [get_node_id] Local node name: srvnode-2.data.private
2021-07-21 07:57:50 alert_monitor INFO [process_alert] Last Nodes IDs: 3
2021-07-21 07:57:50 alert_monitor INFO [process_alert] This node does not have highest id. Local node id : 2
, all nodes: ['1', '2', '3'].
  </code>
</pre>

### CODE REVIEW

**Comment by indrajitzagade** [comment]
@shindeam Please add unit testing details.

File: ha/alert/node_alert_monitor.py [hunk 1]
```diff
@@ -32,4 +33,39 @@ def process_alert(self):
...
+        """
+        self.process = SimpleCommand()
+    
+    def get_status(self):
+        self._nodeids, self._err, self._rc = self.process.run_cmd("pcs status corosync | awk '{print $1}' | tail -n+5")
+        self._local_node, self._err, self._rc = self.process.run_cmd("crm_node -i")
+        Log.info(f"List of online node ids in cluster: {self._nodeids.split()}")
+        Log.info(f"Local node id: {self._local_node}")
+        self.nodeids_list = self._nodeids.split() 
+        self.sort_nodeids(self.nodeids_list, self._local_node)
```
Comment (indrajitzagade) [inline ha/alert/node_alert_monitor.py:61]: Where are we returning two values expected by the caller?

File: ha/alert/node_alert_monitor.py [hunk 2]
```diff
@@ -32,4 +33,39 @@ def process_alert(self):
...
+
+    def sort_nodeids(self, nodeids_list: list, local_node):
+        """
+        Sort node ids in ascending order.
+        """
+        self.nodeids_list = nodeids_list
+        self.local_node = local_node
```
Comment (indrajitzagade) [inline ha/alert/node_alert_monitor.py:68]: Why we need local_node in this method?

File: ha/alert/node_alert_monitor.py [hunk 3]
```diff
@@ -32,4 +33,39 @@ def process_alert(self):
         Log.debug("Processing event for NodeAlertMonitor")
         # Environment variable are avilable in self.crm_env
         self.iem = IemGenerator()
-        self.iem.generate_iem(self.crm_env["CRM_alert_node"], self.alert_event_module, self.alert_event_type)
+        alert_dup = AlertDuplication()
+        self.node_ids, self.local_node = alert_dup.AlertDuplication().get_status()
+
+        if self.node_ids[-1] == self.local_node:
```
Comment (indrajitzagade) [inline ha/alert/node_alert_monitor.py:39]: @shindeam Will this create any timing issue where the IEM will be missed from all nodes?

**Review by indrajitzagade — CHANGES_REQUESTED** [review]
Please fix/resolve the review comments.

**Comment by indrajitzagade** [comment]
@shindeam Please let me know when a final patch after unit testing is ready for the review. Please update the unit testing details in the PR template then too.

File: ha/alert/node_alert_monitor.py [hunk 4]
```diff
@@ -32,4 +32,33 @@ def process_alert(self):
...
+        self.alert_dup = HandleAlertDuplication()
+        self.nodes_ids, self.local_node_id, self.local_node_name = self.alert_dup.get_node_id()
+        # Generate and send IEM only through the highest online node in cluster.
+        if self.nodes_ids[-1] == self.local_node_id:
+            self.iem.generate_iem(self.crm_env["CRM_alert_node"], self.alert_event_module, self.alert_event_type)
+            Log.info(f"Sent IEM alert from local node: {self.local_node_name} / {self.local_node_id}")
+        else:
+            Log.info("Not the highest node to sent IEM node failure alert.")
+
+class HandleAlertDuplication:
```
Comment (ajaysrivas) [inline ha/alert/node_alert_monitor.py:44]: This class is not required. Add _get_all_nodes and _get_local_node in previous class.

File: ha/alert/node_alert_monitor.py [hunk 5]
```diff
@@ -32,4 +32,33 @@ def process_alert(self):
         Log.debug("Processing event for NodeAlertMonitor")
         # Environment variable are avilable in self.crm_env
         self.iem = IemGenerator()
-        self.iem.generate_iem(self.crm_env["CRM_alert_node"], self.alert_event_module, self.alert_event_type)
+        self.alert_dup = HandleAlertDuplication()
+        self.nodes_ids, self.local_node_id, self.local_node_name = self.alert_dup.get_node_id()
+        # Generate and send IEM only through the highest online node in cluster.
+        if self.nodes_ids[-1] == self.local_node_id:
+            self.iem.generate_iem(self.crm_env["CRM_alert_node"], self.alert_event_module, self.alert_event_type)
+            Log.info(f"Sent IEM alert from local node: {self.local_node_name} / {self.local_node_id}")
```
Comment (ajaysrivas) [inline ha/alert/node_alert_monitor.py:40]: Sent IEM alert from the node - name: {self.local_node_name}, id:  {self.local_node_id}

File: ha/alert/node_alert_monitor.py [hunk 6]
```diff
@@ -32,4 +32,33 @@ def process_alert(self):
...
+        else:
+            Log.info("Not the highest node to sent IEM node failure alert.")
```
Comment (ajaysrivas) [inline ha/alert/node_alert_monitor.py:42]: It should be debug.
This node does not have highest id. Local node id : {}, all nodes: {}.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: ha/alert/node_alert_monitor.py (hunk 5 above)
Comment (shindeam) [reply ha/alert/node_alert_monitor.py]: Done.

File: ha/alert/node_alert_monitor.py (hunk 6 above)
Comment (shindeam) [reply ha/alert/node_alert_monitor.py]: Done.

**Pull request closed without merging**